SPEC.md

```diff
@@ -0,0 +1,233 @@
+
+- Tens of AKS nodes; certainly not thousands.
+- Thousands of users, not millions.
+- Full state is tens of megabytes, not gigabytes
+- AKS nodes are incentivized. We assume most of them will do their best to stay available and do the work.
+- Permissionless. AKS nodes will come and go willy nilly and there is no central authority that knows who they are or how reliable they are.
+
+## Secret data CRUD
+
+Anyone on the internet can query the blockchain for the public key and hostname of an AKS node. Using this information they can establish a secure connection to the AKS node. The untrusted host running the CVM cannot MITM this connection.
+
+_TODO: Spell out the reasons why this is secure and how it works in detail, e.g. how the CVM attestation is verified by the user, why this on-chain pubkey registry is sane and what the difference is between this setup and standard TLS certs. I guess the deeper question is: what does the blockchain do for us here?_
+
+~~The blockchain has privileged access (encrypted and authenticated) to a small set of RPC endpoints that allows validator nodes to send secrets, e.g. API keys, on user's behalf to the CVMs running the `entropy-aks` daemon ("AKS" for short). The secret data submitted by users is encrypted in-flight and opaque to validators and the on-chain transaction data cannot be decrypted/inspected. Secrets only ever reside inside the CVM, running in the TDX enclave.~~
+
+Users submit and query data to/from the AKS nodes using standard http requests with encrypted payloads.
+
+_TODO: HTTP requests? Or websockets? Or either? The untrusted cloud operator can inspect the traffic of course, and can spy on metdata (like user's IP). Or how does this work? Is the client-to-AKS transport mechanism specc'd somewhere?_
+
+The relevant RPC endpoints exposed by the AKS are:
+
+- `POST /secret?data=&acl=`: adds a new secret with an appropriate ACL; returns a `secret_id`.
+- `PUT /secret?id=&data=&acl`: updates the secret and/or ACL with secret_id `id`.
+- `DELETE /secret?id`: deletes the secret with secret_id `id`.
+- `GET /make-request?id=&remote-url=&verb=&payload=`: instructs the AKS to make a remote API call to `remote-url` using the http verb `verb` and the secret matching the `id` to send the payload in `payload`.
+
+*NOTE*: `secret_id`s should be unique and **random**, given that knowledge of a `secret_id` allows anyone to make remote API calls on behalf of the user.
+
+_TODO: Use correct endpoints, from the spec doc._
+
+_TODO: Describe the ACL format and how it works in detail._
+
+_TODO: Can users provide any `remote-url` like this or is it limited to a set of whitelisted URLs? Stored where? In the ACL? Or separate?_
+
+~~All calls to the AKS from the outside are signed and contain counter measures to replay attacks (likely a block number and/or a random nonce). The API responses are encrypted to the end user's public key before they are send back outside the AKS.~~
```

Comment:
Currently we are using timestamp checks.  Which means we rely on client and sever to have accurate clocks (which these days is not so unreasonable to assume).

In `entropy-tss` we usually use block number, and we may well end up doing that here.  But we have had a lot of issues with automated tests and block number checks.

This is an argument for https - the TLS handshake means we don't need to worry about replay attacks.

> The API responses are encrypted to the end user's public key before they are send back outside the AKS.

Currently they are just sent clear.  Again, with https we would not need to implement this.